### PR TITLE
Add SAML IdP plugin for internal use

### PR DESCRIPTION
**Pull request by tvernum**
This change merges the "feature-internal-idp" branch into Elasticsearch.

This introduces a small `identity-provider` plugin as a child of the `x-pack` module.
This allows ES to act as a SAML IdP, for users who are authenticated against the
Elasticsearch cluster.

This feature is intended for internal use within Elastic Cloud environments
and is not supported for any other use case. It falls under an enterprise license tier.

The IdP is disabled by default.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/SerializationTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+/**
+ * Utility methods for assisting with testing {@link Writeable} object.
+ */
+public class SerializationTestUtils {
```
Comment (jasontedor) [inline test/framework/src/main/java/org/elasticsearch/test/SerializationTestUtils.java]: This is a drive-by comment. I can't find any uses of this? And also, I think that `AbstractWireSerializingTestCase` provides roughly the same functionality?

Comment (tvernum) [reply test/framework/src/main/java/org/elasticsearch/test/SerializationTestUtils.java]: It's used in 
- idp.action.DeleteSamlServiceProviderRequestTests
- idp.action.PutSamlServiceProviderRequestTests
- idp.saml.sp.SamlServiceProviderDocumentTests
- idp.saml.support.SamlAuthenticationStateTests

You're right, I can switch those to use the existing methods.
I was sure we had something like that, but I just couldn't find it, so I figured I must have remembered incorrectly.

Comment (jasontedor) [reply test/framework/src/main/java/org/elasticsearch/test/SerializationTestUtils.java]: Strange, the IDE showed me there were not any uses. Either way, thanks for switching over.

**Review by jkakavas — APPROVED** [review]
:shipit: 

(after we merge https://github.com/elastic/elasticsearch/pull/54053 )

**Pull request merged**